### PR TITLE
[WIP]feature: support for `host + uri` as an index (radixtree)

### PR DESCRIPTION
**Pull request by bobopan**
Fix https://github.com/iresty/apisix/issues/654

### CODE REVIEW

File: lua/apisix/http/router/radixtree_host_uri.lua [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+local type = type
+local str_reverse = string.reverse
+local error = error
+local user_routes
+local cached_version
+
+
+local _M = {version = 0.1}
+
+    local only_uri_routes = {}
```
Comment (membphis) [inline lua/apisix/http/router/radixtree_host_uri.lua]: `only_uri_routes`, `only_uri_router `, `host_uri_routes ` and `host_uri_router` shoud be a top level variable.

File: lua/apisix/http/router/radixtree_host_uri.lua [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    for _, route in ipairs(routes) do
+        if type(route) == "table" then
+            core.table.insert(only_uri_routes, {
+                paths = route.uris or route.uri,
+                handler = route.handler,
+                method = route.methods,
+            })
+        end
+    end
+    only_uri_routes = only_uri_routes or {}
```
Comment (membphis) [inline lua/apisix/http/router/radixtree_host_uri.lua]: we can remove this line, it's useless

File: lua/apisix/http/router/radixtree_host_uri.lua [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        end
+    end
+    return
+end
+
+
+local function create_radixtree_router(routes)
+    routes = routes or {}
+
+    local api_routes = plugin.api_routes()
```
Comment (membphis) [inline lua/apisix/http/router/radixtree_host_uri.lua]: we can remove this line too(useless)

File: lua/apisix/http/router/radixtree_host_uri.lua [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    core.table.clear(host_uri_routes)
+    core.table.clear(only_uri_routes)
+    for _, route in ipairs(routes or {}) do
+        push_radixtree_host_router(route)
+    end
+
+    create_only_uri_router()
+
+    core.log.error("route items: ",
```
Comment (membphis) [inline lua/apisix/http/router/radixtree_host_uri.lua]: that should be a debug log, so we'd better use `info` level.

File: lua/apisix/http/router/radixtree_host_uri.lua [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    end
+    local host = route.value.host
+    local hosts = route.value.hosts
+    local uri = route.value.uris or route.value.uri
+    if (not host or type(host) ~= "string") and (not hosts or type(hosts) ~= "table") then
+        core.table.insert(only_uri_routes, {
+            paths = route.value.uris or route.value.uri,
+            method = route.value.methods,
+            hosts = hosts or host,
+            remote_addrs = route.value.remote_addr,
```
Comment (membphis) [inline lua/apisix/http/router/radixtree_host_uri.lua]: `route.value.remote_addr or route.value.remote_addrs`

File: lua/apisix/http/router/radixtree_host_uri.lua [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        end
+    })
+end
+
+
+local function push_radixtree_host_router(route)
+    if type(route) ~= "table" then
+        return
+    end
+    local host = route.value.host
```
Comment (membphis) [inline lua/apisix/http/router/radixtree_host_uri.lua]: code style: add a blank line

**Comment by bobopan** [comment]
close

**Pull request closed without merging**